checker/src/org/checkerframework/checker/lock/wholefiles/LockVisitor.java

```diff
@@ -0,0 +1,580 @@
+    private final Class<? extends Annotation> checkerGuardSatisfiedClass = GuardSatisfied.class;
+    private final Class<? extends Annotation> checkerLockHeldClass = LockHeld.class;
+
+    // Note that Javax and JCIP @GuardedBy is used on both methods and objects. For methods they are
+    // equivalent to the Checker Framework @Holding annotation.
+    private final Class<? extends Annotation> javaxGuardedByClass = javax.annotation.concurrent.GuardedBy.class;
+    private final Class<? extends Annotation> jcipGuardedByClass = net.jcip.annotations.GuardedBy.class;
+
+    /** Annotation constants */
+    protected final AnnotationMirror GUARDEDBY, GUARDEDBYINACCESSIBLE, GUARDSATISFIED, GUARDEDBYBOTTOM;
+
+    public LockVisitor(BaseTypeChecker checker) {
+        super(checker);
+
+        GUARDEDBYINACCESSIBLE = AnnotationUtils.fromClass(elements, GuardedByInaccessible.class);
+        GUARDEDBY = AnnotationUtils.fromClass(elements, GuardedBy.class);
+        GUARDSATISFIED = AnnotationUtils.fromClass(elements, GuardSatisfied.class);
+        GUARDEDBYBOTTOM = AnnotationUtils.fromClass(elements, GuardedByBottom.class);
+
+        checkForAnnotatedJdk();
+    }
+
+    @Override
+    public Void visitVariable(VariableTree node, Void p) { // visit a variable declaration
+        // A user may not annotate a primitive type with any qualifier from the @GuardedBy hierarchy.
+        if (node.getType().getKind() == Kind.PRIMITIVE_TYPE &&
+            (node.toString().contains("GuardSatisfied") ||
+             node.toString().contains("GuardedBy"))){ // HACK!!! TODO: Fix once there is a way to reliably retrieve user-written qualifiers.
+            checker.report(Result.failure("primitive.type.guardedby"), node);
+        }
+        return super.visitVariable(node, p);
+    }
+
+    @Override
+    public LockAnnotatedTypeFactory createTypeFactory() {
+        // We need to directly access useFlow from the checker, because this method gets called
+        // by the superclass constructor and a field in this class would not be initialized
+        // yet. Oh the pain.
+        return new LockAnnotatedTypeFactory(checker, true);
+    }
+
+    // Issue an error if a method (explicitly or implicitly) annotated with @MayReleaseLocks has a formal parameter
+    // or receiver (explicitly or implicitly) annotated with @GuardSatisfied.
+    @Override
+    public Void visitMethod(MethodTree node, Void p) {
+
+        SideEffectAnnotation sea = atypeFactory.methodSideEffectAnnotation(TreeUtils.elementFromDeclaration(node), true);
+
+        if (sea == SideEffectAnnotation.MAYRELEASELOCKS) {
+            boolean issueGSwithMRLWarning = false;
+
+            VariableTree receiver = node.getReceiverParameter();
+            if (receiver != null) {
+                if (AnnotationUtils.areSameByClass(atypeFactory.getAnnotatedType(receiver).getAnnotationInHierarchy(GUARDEDBYINACCESSIBLE),
```

Comment:
Could be replaced with:
atypeFactory.getAnnotatedType(receiver).hasAnnotation(checkerGuardSatisfiedClass)